horovod/mxnet/partition_trainer/pos_trainer.py

```diff
@@ -0,0 +1,174 @@
+        param2rank = {}
+        sizes = [0 for _ in range(world_size)]
+        for param in params:
+            if param.grad_req != 'null':
+                current_rank = sizes.index(min(sizes))
+                partition_params[current_rank].append(param)
+                num = 1
+                param2rank[param._uuid] = current_rank
+                for p in param.shape:
+                    num *= p
+                sizes[current_rank] += num
+        return partition_params, param2rank
+
+    def _allreduce_grads(self):
+        for i, param in enumerate(self._all_params):
+            if param.grad_req != 'null':
+                allreduce_(param.list_grad()[0], average=False,
+                           name=self._prefix + str(i), priority=-i,
+                           prescale_factor=1.0 / self._gradient_predivide_factor)
+
+    def step(self, batch_size, ignore_stale_grad=False):
+        """
+        inherit from trainer, only call boardcast to make sure all parameter are consistent
+        Makes one step of parameter update. Should be called after
+        `autograd.backward()` and outside of `record()` scope.
+        For normal parameter updates, `step()` should be used, which internally calls
+        `allreduce_grads()` and then `update()`. However, if you need to get the reduced
+        gradients to perform certain transformation, such as in gradient clipping, then
+        you may want to manually call `allreduce_grads()` and `update()` separately.
+        Parameters
+        ----------
+        batch_size : int
+            Batch size of data processed. Gradient will be normalized by `1/batch_size`.
+            Set this to 1 if you normalized loss manually with `loss = mean(loss)`.
+        ignore_stale_grad : bool, optional, default=False
+            If true, ignores Parameters with stale gradient (gradient that has not
+            been updated by `backward` after last step) and skip update.
+        """
+        rescale_grad = self._scale / batch_size
+        self._check_and_rescale_grad(rescale_grad)
+
+        if not self._kv_initialized:
+            self._init_kvstore()
+        if self._params_to_init:
+            self._init_params()
+
+        self._allreduce_grads()
+        self._update(ignore_stale_grad)
+        self.broadcast_params()
+
+    def update(self, batch_size, ignore_stale_grad=False):
+        if not self._kv_initialized:
+            self._init_kvstore()
+        if self._params_to_init:
+            self._init_params()
+        assert not (self._kvstore and self._update_on_kvstore), \
+            'update() when parameters are updated on kvstore ' \
+            'is not supported. Try setting `update_on_kvstore` ' \
+            'to False when creating trainer.'
+
+        self._check_and_rescale_grad(self._scale / batch_size)
+        self._update(ignore_stale_grad)
+        self.broadcast_params()
+
+
+
+    def broadcast_params(self):

```

Comment:
This should also be renamed to something like `_broadcast_partitioned_params` to better differentiate it from the existing `_broadcast_parameters` method. Also, add a brief comment describing what it does.